pkg/controlplane/controller/leaderelection/leaderelection_controller.go

```diff
@@ -0,0 +1,660 @@
+			c.enqueueCandidate(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			c.enqueueCandidate(newObj)
+		},
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	c.leaseSynced = leaseSynced.HasSynced
+	c.leaseCandidateSynced = leaseCandidateSynced.HasSynced
+	return c, nil
+}
+
+func (c *Controller) enqueueLease(obj any) {
+	if lease, ok := obj.(*v1.Lease); ok {
+		key, err := controller.KeyFunc(lease)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("cannot get name of object %v: %w", lease, err))
+		}
+		c.leaseQueue.Add(key)
+	}
+}
+
+func (c *Controller) enqueueCandidate(obj any) {
+	if lease, ok := obj.(*v1alpha1.LeaseCandidate); ok {
+		_ = c.reconcileLeaseCandidate(lease)
+	}
+}
+
+func (c *Controller) runWorker(ctx context.Context) {
+	klog.Infof("runWorker")
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *Controller) runElectionWorker(ctx context.Context) {
+	klog.Infof("runElectionWorker")
+	for c.processNextElectionItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextElectionItem(ctx context.Context) bool {
+	key, shutdown := c.electionQueue.Get()
+	if shutdown {
+		return false
+	}
+	defer c.electionQueue.Done(key)
+
+	err := c.runElection(ctx, key)
+	utilruntime.HandleError(err)
+
+	return true
+}
+
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	key, shutdown := c.leaseQueue.Get()
+	if shutdown {
+		return false
+	}
+	defer c.leaseQueue.Done(key)
+
+	err := func() error {
+		return c.reconcile(ctx, key)
+	}()
+
+	if err == nil {
+		c.leaseQueue.AddAfter(key, requeueInterval)
+		return true
+	}
+
+	utilruntime.HandleError(err)
+	klog.Infof("processNextWorkItem.AddRateLimited: %v", key)
+	c.leaseQueue.AddRateLimited(key)
+
+	return true
+}
+
+func (c *Controller) reconcileLeaseCandidate(lc *v1alpha1.LeaseCandidate) error {
+	if lc == nil {
+		return nil
+	}
+	klog.Infof("reconcile found canLead label namespace=%q, name=%q: %q", lc.Namespace, lc.Name, lc.Spec.LeaseName)
+	leaderLeaseID := leaderLeaseID{lc.Namespace, lc.Spec.LeaseName}
+
+	// Ignore candidates that transitioned to Pending because reelection is already in progress
+	if lc.Spec.PingTime != nil {
+		return nil
+	}
+
+	// Updates during election will be marked as ack
+	status := c.getLS(leaderLeaseID)
+	if status.candidatesAck != nil {
+		status.candidateAck(lc)
+		return nil
+	}
+	// The current discriminating criteria is inElection vs not in election.
+	// Not all updates require election and more short circuiting logic would be beneficial.
+	c.electionQueue.Add(leaderLeaseID)
+	return nil
+
+	// klog.Infof("Adding lease %s", leaderLeaseID.String())
+	// c.leaseQueue.Add(leaderLeaseID.String())
+	// return nil
+}
+
+func (c *Controller) reconcile(ctx context.Context, key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	lease, err := c.leaseInformer.Lister().Leases(namespace).Get(name)
+
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
```

Comment:
```suggestion
	lease, err := c.leaseInformer.Lister().Leases(namespace).Get(name)
	if err != nil && !apierrors.IsNotFound(err) {
		return err
	} else if apierrors.IsNotFound(err) {
		return nil
	}
```

Reply:
I'll do a better job documenting, but if lease is notfound we need to keep going. It's possible that the lease was just deleted (eg: leader went away) but candidates are still available.

Reply:
this code is processing the lease from the event handler, so I don't think you should go to the informer to get the Lease from there and just use the Lease you are getting as input parameter